dali/util/uri.cc

```diff
@@ -0,0 +1,171 @@
+// Copyright (c) 2024, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "dali/util/uri.h"
+#include <utility>
+#include <stdexcept>
+
+namespace dali {
+
+inline bool allowed_scheme_char(char c) {
+  return std::isalnum(c) || c == '.' || c == '+' || c == '-';
+}
+
+inline bool allowed_char(char c) {
+  // See https://en.wikipedia.org/wiki/Uniform_Resource_Identifier
+  // gen-delims: :, /, ?, #, [, ], and @
+  // sub-delims: !, $, &, ', (, ), *, +, ,, ;, and =
+  // unreserved characters (uppercase and lowercase letters, decimal digits, -, ., _, and ~)
+  // the character %
+  static const std::string gen_delims = ":/?#[]@";
+  static const std::string sub_delims = "!$&'()*+,;=";
+  static const std::string unreserved = "-._~";
+  return (std::isalnum(c)
+    || gen_delims.find(c) != std::string::npos
+    || sub_delims.find(c) != std::string::npos
+    || unreserved.find(c) != std::string::npos);
+}
+
+std::string display_char(char c) {
+  switch (c) {
+    case '\a':
+      return "\\a";
+    case '\b':
+      return "\\b";
+    case '\t':
+      return "\\t";
+    case '\n':
+      return "\\n";
+    case '\v':
+      return "\\v";
+    case '\f':
+      return "\\f";
+    case '\r':
+      return "\\r";
+    case '\"':
+      return "\\\"";
+    case '\'':
+      return "\\\'";
+    case '\?':
+      return "\\\?";
+    case '\\':
+      return "\\\\";
```

Comment:
I don't think we need to escape the following: `' " \` - they are perfectly displayable.